llvm/lib/MCCAS/MCCASObjectV1.cpp

```diff
@@ -3155,25 +3166,153 @@ static AbbrevEntryReader getAbbrevEntryReader(ArrayRef<StringRef> AbbrevEntries,
+  }
+  case dwarf::DW_FORM_ref_sig8:
+  case dwarf::DW_FORM_data8:
+  case dwarf::DW_FORM_ref8:
+  case dwarf::DW_FORM_ref_sup8: {
+    return std::make_pair(false, 8);
+  }
+  case dwarf::DW_FORM_data16: {
+    return std::make_pair(false, 16);
+  }
+
+  case dwarf::DW_FORM_strp:
+  case dwarf::DW_FORM_sec_offset:
+  case dwarf::DW_FORM_GNU_ref_alt:
+  case dwarf::DW_FORM_GNU_strp_alt:
+  case dwarf::DW_FORM_line_strp:
+  case dwarf::DW_FORM_strp_sup: {
+    auto FormSize = FP.getDwarfOffsetByteSize();
+    return std::make_pair(false, FormSize);
+  }
+  case dwarf::DW_FORM_addrx3:
+  case dwarf::DW_FORM_lo_user: {
+    llvm_unreachable("usupported form");
+    break;
+  }
+  }
+}
+
+Error DIEVisitor::materializeAbbrevDIE(unsigned AbbrevIdx) {
+  constexpr auto AddrSize = 8;
+  constexpr auto FormParams =
+      dwarf::FormParams{4 /*Version*/, AddrSize, dwarf::DwarfFormat::DWARF32};
+
   AbbrevEntryReader AbbrevReader =
       getAbbrevEntryReader(AbbrevEntries, AbbrevIdx);
 
-  if (Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag())
-    StartTagCallback(*MaybeTag, AbbrevIdx);
-  else
+  Expected<dwarf::Tag> MaybeTag = AbbrevReader.readTag();
+  if (!MaybeTag)
     return MaybeTag.takeError();
 
   Expected<bool> MaybeHasChildren = AbbrevReader.readHasChildren();
   if (!MaybeHasChildren)
     return MaybeHasChildren.takeError();
 
-  if (auto E = visitDIEAttrs(AbbrevReader, DataReader, DIEData))
+  AbbrevDIETagAndChildren TagAndChildren;
+  TagAndChildren.Tag = *MaybeTag;
+  TagAndChildren.HasChildren = *MaybeHasChildren;
+
+  SmallVector<AbbrevDIEContents> AbbrevVector;
+  while (true) {
+    AbbrevDIEContents Contents;
+    Expected<dwarf::Attribute> Attr = AbbrevReader.readAttr();
+    if (!Attr)
+      return Attr.takeError();
+    if (*Attr == getEndOfAttributesMarker())
+      break;
+
+    Expected<dwarf::Form> Form = AbbrevReader.readForm();
+    if (!Form)
+      return Form.takeError();
+    Contents.Attr = *Attr;
+    Contents.Form = *Form;
+    Contents.FormInDistinctData = doesntDedup(*Form, *Attr);
+    auto FormSizeInfo = getNonULEBFormSize(*Form, FormParams);
+    Contents.FormSizeRequiresDIEContents = FormSizeInfo.first;
+    Contents.FormSize = FormSizeInfo.second;
+    AbbrevVector.push_back(Contents);
+  }
+  AbbrevDIECache.push_back(std::make_pair(TagAndChildren, AbbrevVector));
+  return Error::success();
+}
+
+Error DIEVisitor::visitDIERef(BinaryStreamReader &DataReader,
+                              unsigned AbbrevIdx, StringRef DIEData,
+                              ArrayRef<DIEDataRef> &DIEChildrenStack) {
+
+  AbbrevDIECache.reserve(AbbrevEntries.size());
+  for (unsigned I = 0; I < AbbrevEntries.size(); I++) {
+    if (Error E = materializeAbbrevDIE(encodeAbbrevIndex(I)))
+      return E;
+  }
+
+  auto TagAndChildren =
+      AbbrevDIECache[decodeAbbrevIndexAsAbbrevSetIdx(AbbrevIdx)].first;
+  StartTagCallback(TagAndChildren.Tag, AbbrevIdx);
+
+  if (auto E = visitDIEAttrs(DataReader, DIEData, AbbrevIdx))
```

Comment:
Let's avoid querying the cache twice, we can replace `AbbrevIdx` here with the cached entry (its `.second` member)